comet/redis_async_locks.py

```diff
@@ -17,172 +23,260 @@ def __init__(self, message: str):
 
 
 class Lock:
-    """Async context manager for redis lock."""
+    """Async context manager for redis lock.
+
+    This lock has a similar API to `asyncio.Lock`. The main difference is
+    that acquiring the lock returns an exclusive `redis` connection for use
+    by the locking task. As this is an open and active connection it can and
+    should be used within the locked region to prevent other tasks starving
+    the locking task of connections preventing it from unlocking the region.
+
+    The lock requires a minimum of one connection to function.
+
+    Parameters
+    ----------
+    redis : aioredis.ConnectionsPool
+        A connections pool instance that will be used to connect to the redis database.
+    name : str
+        A name for the lock. This must be unique (i.e. not clash with
+        other locks), and be set across all processes that want to use
+        the same lock.
+    """
 
     def __init__(self, redis, name):
-        """Create context manager."""
+        """Create context manager.
+        """
         self.name = name
         self.redis = redis
+        self._redis_conn = None
 
-    async def __aenter__(self):
-        """Acquire lock."""
-        await redis_lock_acquire(self.redis, self.name)
+    @classmethod
+    async def create(cls, redis, name):
 
-    async def __aexit__(self, exc_type, exc, tb):
-        """Release lock."""
-        await redis_lock_release(self.redis, self.name)
+        self = cls(redis, name)
 
+        # Clear the lock and recreate
```

Comment:
I think the answer is no to this. I looked into it, and the problem is that you can't await co-routines because the constructor is synchronous. In some cases you can get around it by getting the event loop and running a coroutine from within the constructor, but that doesn't work here because you need to be able to create the lock in asynchronous code.

Recommendations from the internet are basically to do exactly this, have a `create` factory method (https://stackoverflow.com/questions/33128325/how-to-set-class-attribute-with-await-in-init). That link does mention a few hacks to make constructors async, but I think they're best avoided.

Reply:
One solution might be using a synchronous redis client for initializing the lock. But this is fine as is in my opinion.